platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,493 @@
+The Roster, a subset of the address book, will provide the data structure and API for this mechanism.
+This proposal provides a specification for the behavior of rosters starting from their creation from the Address Book to
+their terminal state.
+A roster reaches a terminal state when it is either adopted by the platform or replaced by a new roster.
+The candidate roster will always get adopted on the next software upgrade.
+
+The Hedera App (henceforth reffered to as 'App') will maintain a version of the Address Book as a dynamic list that
+reflects the desired future state of
+the
+network's nodes. This dynamic list of Addresses will be continuously updated by HAPI transactions, such as those that
+create, update, or delete nodes.
+At some point when the App decides it's time to begin work on adopting a new address book (scope beyond this proposal),
+it will create a `Roster` object with information from the Address Book and pass it to the platform.
+
+The mechanism for doing so is detailed below.
+
+## What this proposal does not address
+
+## Requirements
+
+- Immutability: The roster hash will be a critical piece of data. Rosters must be immutable to protect the integrity
+  of its computed hash.
+- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
+  copying.
+- Clear API: A well-defined API should be provided for the App to create and submit candidate rosters to the platform,
+  and for the
+  platform to manage the lifecycle of submitted rosters.
+- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
+  Event Validation etc.) must be adapted to use the new `Roster` object instead.
+- Address book paradigm discontinued within the platform codebase in favor of `Roster`s
+
+## Architecture
+
+###### Roster Lifecycle
+
+![](TSS%20Roster%20Lifecycle.drawio.svg)
+
+### Roster Public API
+
+#### Setting the Candidate Roster
+
+A new method will be added to the platform state to allow the App to submit a Candidate Roster to the platform:
+
+```java
+//in PlaformState
+
+void setCandidateRoster(@NonNull final Roster candidateRoster);
+```
+
+The steps to validate and add the Roster to the state will be executed within the same thread that submits the request.
+An exception will be thrown if the Roster is invalid.
+The durability of the submitted Roster is ensured via storing it in the state.
+
+The existing `PlatformState` class is in the middle of a refactor, and in the long term, may cease to exist.
+However, its replacement will continue to have this new API to set a candidate roster.
+
+## Data Structure
+
+The App is already responsible for managing the address book.
+We propose that it continues to do so, and when it receives a HAPI transaction, creates a candidate Roster object
+and passes it to the platform via the new API. The platform will then validate the roster and store it in the state.
```

Comment:
In reality a new candidate roster will be generated at most once per day (first transaction after midnight). In the end, it is up to the app to decide when to propose a new roster.